pulsar-client/src/main/java/org/apache/pulsar/client/impl/SinglePartitionMessageRouterImpl.java

```diff
@@ -18,26 +18,37 @@
  */
 package org.apache.pulsar.client.impl;
 
+import java.nio.charset.StandardCharsets;
 import java.util.Random;
 
+import com.google.common.hash.HashCode;
+import com.google.common.hash.HashFunction;
+import com.google.common.hash.Hashing;
 import org.apache.pulsar.client.api.Message;
 import org.apache.pulsar.client.api.MessageRouter;
 
-public class SinglePartitionMessageRouterImpl implements MessageRouter {
+public class SinglePartitionMessageRouterImpl extends MessageRouterBase {
 
     private final int partitionIndex;
-    private final int numPartitions;
 
-    public SinglePartitionMessageRouterImpl(int numPartitions) {
+    public SinglePartitionMessageRouterImpl(int numPartitions, boolean useMurmurHash) {
+        super(numPartitions, useMurmurHash);
         this.partitionIndex = new Random().nextInt(numPartitions);
-        this.numPartitions = numPartitions;
     }
 
     @Override
     public int choosePartition(Message msg) {
         // If the message has a key, it supersedes the single partition routing policy
         if (msg.hasKey()) {
-            return ((msg.getKey().hashCode() & Integer.MAX_VALUE) % numPartitions);
+            if (useMurmurHash) {
+                HashFunction hf = Hashing.murmur3_32(0);
```

Comment:
I think this logic here is not allocation-friendly. It generates two objects `HashFunction` and `HashCode` per message here.

I would suggest using a MurmurHash class that is a static method take takes a bytes array or a string to complete a hash value, rather than using guava `HashFunction` or `Hasher`. E.g.

```

public class MurmurHash3 {

     public static int hash32(ByteBuf data, int offset, int length, int seed) { ... }

}
```

Another suggestion - it would be good to write a jmh microbenchmark, testing the difference between using MurmurHash3 and simple hash.

Reply:
@sijie I found the implementation of MurmurHash3 into Pulsar repository!
https://github.com/apache/incubator-pulsar/blob/554f0505fc424896d820d639252b28e0ec523526/pulsar-checksum/src/main/java/com/scurrilous/circe/CommonHashes.java